File: src/ua/edu/odeku/ceem/mapRadar/tools/importGeoName/GeoNameImporter.java
package ua.edu.odeku.ceem.mapRadar.tools.importGeoName;

import org.hibernate.Session;
import ua.edu.odeku.ceem.mapRadar.db.DB;
import ua.edu.odeku.ceem.mapRadar.db.models.GeoName;
import ua.edu.odeku.ceem.mapRadar.exceptions.db.models.GeoNameException;
import ua.edu.odeku.ceem.mapRadar.utils.gui.UserMessage;
import ua.edu.odeku.ceem.mapRadar.utils.thread.Handler;

import javax.persistence.EntityManager;
import javax.swing.*;
import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.IOException;
import java.sql.SQLException;
import java.util.LinkedList;

/**
 * Файл поток, занимает импортом данных из файла в базу данных
 * User: Aleo skype: aleo72
 * Date: 10.11.13
 * Time: 18:16
 */
public class GeoNameImporter extends Thread {

    public boolean stop = false;
    private File file;
    private JProgressBar progressBar;
    private String[] lines;
    private Handler handler;

    public GeoNameImporter(JProgressBar progressBar) {
        this.progressBar = progressBar;
    }

    public GeoNameImporter(JProgressBar progressBar, Handler handler) {
        this.progressBar = progressBar;
        this.handler = handler;
    }

    public void setFileInput(File file) {
        this.file = file;
    }

    @Override
    public void run() {
        try{
            if (!stop && file != null) {
                progressBar.setValue(0);
                lines = readFile();
                Session session = DB.getSession();

                try {
                    if (lines != null && lines.length > 0) {
                        progressBar.setMaximum(lines.length);
                        for (String line : lines) {
                            if(stop)
                                break;
                            progressBar.setValue(progressBar.getValue() + 1);
                            GeoName geoName = GeoName.createGeoName(line);
                            try {
                                session.save(geoName);
                            }
                            catch (Exception e){
                                System.err.println(e.getMessage());
                            }
                        }

                    }
                } catch (GeoNameException e) {
                    e.printStackTrace();
                    UserMessage.error(null, e.getMessage());
                } finally {
                    DB.closeSession(session);
                }

            }
        } catch (Exception ex){
            ex.printStackTrace();
        } finally {
            clear();
            if(handler != null){
                handler.start();
            }
        }
    }

    private void clear() {
        stop = false;
        progressBar.setValue(0);
        file = null;
        lines = null;
    }

    private String[] readFile() {
        String[] lines = null;
        try (BufferedReader br = new BufferedReader(new FileReader(file))) {
            String line;
            LinkedList<String> list = new LinkedList<String>();
            while ((line = br.readLine()) != null && !stop) {
                list.add(line);
            }
            lines = list.toArray(new String[0]);
        } catch (IOException e) {
            e.printStackTrace();
            UserMessage.error(null, e.getMessage());
        }

        if (stop)
            lines = null;

        return lines;
    }

}

File: src/ua/edu/odeku/ceem/mapRadar/tools/importGeoName/GeoNameImporter.scala
/*
 * Odessa State environmental University
 * Copyright (C) 2013
 */

package ua.edu.odeku.ceem.mapRadar.tools.importGeoName

import java.io.{IOException, File}
import javax.swing.JProgressBar
import ua.edu.odeku.ceem.mapRadar.utils.thread.Handler
import ua.edu.odeku.ceem.mapRadar.db.DB
import ua.edu.odeku.ceem.mapRadar.db.model.{GeoNames, GeoName}
import ua.edu.odeku.ceem.mapRadar.utils.gui.UserMessage
import ua.edu.odeku.ceem.mapRadar.tools.ToolFrame
import ua.edu.odeku.ceem.mapRadar.exceptions.db.models.GeoNameException
import java.util

/**
 * User: Aleo Bakalov
 * Date: 11.12.13
 * Time: 13:34
 */
class GeoNameImporter(val progressBar: JProgressBar, val closeHandler: Function0[Unit]) extends Thread {

	var stopFlag: Boolean = false
	private var file: File = null
	private var lines: Array[String] = null

	def this(progressBar: JProgressBar) {
		this(progressBar, () => {})
	}

	def setFileInput(file: File) {
		this.file = file
	}

	override def run() {
		try {
			if (!stopFlag && file != null) {
				progressBar.setValue(0)
				lines = readFile
				try {
					if (lines != null && lines.length > 0) {

						progressBar.setMaximum(lines.length)

						DB.database withSession {
							implicit session =>
								for (line <- lines; if !stopFlag) {

									progressBar.setValue(progressBar.getValue + 1)
									val geoName = GeoNames.createGeoName(line)

									try {
										GeoNames += geoName
									}
									catch {
										case e: Exception => System.err.println(e.getMessage)
									}
								}
						}
					}
				}
				catch {
					case e: GeoNameException =>
						e.printStackTrace()
						UserMessage.error(null, e.getMessage)
				}
			}
		}
		catch {
			case ex: Exception => ex.printStackTrace()
		}
		finally {
			clear()
			closeHandler.apply()
		}
	}

	private def clear() {
		stopFlag = false
		progressBar.setValue(0)
		file = null
		lines = null
	}

	private def readFile: Array[String] = {
		var lines: Array[String] = null
		try {
			var line: String = null
			val list: util.LinkedList[String] = new util.LinkedList[String]
			val iter = scala.io.Source.fromFile(file).getLines()
			for (line <- iter; if !line.trim.isEmpty; if !stopFlag) {
				list add line
			}
			lines = list.toArray(new Array[String](0))
		}
		catch {
			case e: IOException =>
				e.printStackTrace()
				UserMessage.error(null, e.getMessage)
		}
		if (stopFlag) lines = null
		lines
	}
}
